src/topics/scheduled.js

```diff
@@ -0,0 +1,108 @@
+	topicsData = topicsData.filter(topicData => Boolean(topicData));
+	const uids = topicsData
+		.map(topicData => topicData.uid)
+		// Filter guests topics
+		.filter(uid => uid)
+		.filter((uid, idx, arr) => arr.indexOf(uid) === idx);
+
+	await Promise.all([].concat(
+		sendNotifications(uids, topicsData),
+		updateUserLastposttimes(uids, topicsData),
+		topicsData.map(topicData => categories.updateRecentTid(topicData.cid, topicData.tid)),
+		topicsData.map(topicData => unpin(topicData.tid, topicData)),
+		updateUnreadCounts(),
+		db.sortedSetsRemoveRangeByScore([`topics:scheduled`], '-inf', now)
+	));
+};
+
+Scheduled.pin = async function (tid, topicData) {
+	return [
+		topics.setTopicField(tid, 'pinned', 1),
+		db.sortedSetAdd(`cid:${topicData.cid}:tids:pinned`, Date.now(), tid),
+		db.sortedSetsRemove([
+			`cid:${topicData.cid}:tids`,
+			`cid:${topicData.cid}:tids:posts`,
+			`cid:${topicData.cid}:tids:votes`,
+		], tid),
+	];
+};
+
+async function unpin(tid, topicData) {
+	return [
+		topics.setTopicField(tid, 'pinned', 0),
+		topics.deleteTopicField(tid, 'pinExpiry'),
+		db.sortedSetRemove(`cid:${topicData.cid}:tids:pinned`, tid),
+		db.sortedSetAddBulk([
+			[`cid:${topicData.cid}:tids`, topicData.lastposttime, tid],
+			[`cid:${topicData.cid}:tids:posts`, topicData.postcount, tid],
+			[`cid:${topicData.cid}:tids:votes`, parseInt(topicData.votes, 10) || 0, tid],
+		]),
+	];
+}
+
+async function sendNotifications(uids, topicsData) {
+	const usernames = await Promise.all(uids.map(uid => user.getUserField(uid, 'username')));
+	const uidToUsername = Object.fromEntries(uids.map((uid, idx) => [uid, usernames[idx]]));
+
+	const postsData = await posts.getPostsData(topicsData.map(({ mainPid }) => mainPid));
+	postsData.forEach((postData, idx) => {
+		postData.user = {};
+		postData.user.username = uidToUsername[postData.uid];
+		postData.topic = topicsData[idx];
+	});
+
+	return topicsData.map(
+		(t, idx) => user.notifications.sendTopicNotificationToFollowers(t.uid, t, postsData[idx])
+	).concat(
+		topicsData.map(
+			(t, idx) => socketHelpers.notifyNew(t.uid, 'newTopic', { posts: [postsData[idx]], topic: t })
+		)
+	);
+}
+
+async function updateUserLastposttimes(uids, topicsData) {
+	const lastposttimes = await Promise.all(uids.map(uid => user.getUserField(uid, 'lastposttime')));
+
+	let timestampByUid = {};
+	topicsData.forEach((tD) => {
+		timestampByUid[tD.uid] = timestampByUid[tD.uid] ? timestampByUid[tD.uid].concat(tD.timestamp) : [tD.timestamp];
+	});
+	timestampByUid = Object.fromEntries(Object.entries(uidTimestamp => [uidTimestamp[0], Math.max(uidTimestamp[1])]));
+
+	const uidsToUpdate = uids.filter((uid, idx) => timestampByUid[uid] > lastposttimes[idx]);
+	return uidsToUpdate.map(uid => user.setUserField(uid, 'lastposttime', timestampByUid[uid]));
+}
+
+async function updateUnreadCounts() {
+	const onlineUids = await user.getUidsFromSet('users:online', 0, -1);
+	return onlineUids.map(uid => topics.pushUnreadCount(uid));
```

Comment:
Take a look at `socketHelpers.notifyNew`. 

It is used to notify online users when a new topic happens. It also does a lot of checks to make sure only the relevant users are notified. https://github.com/NodeBB/NodeBB/blob/master/src/socket.io/helpers.js#L25

It call would be something like

`socketHelpers.notifyNew(uidOfTopicPoster, 'newTopic', { posts: [mainPostData], topic: topicData });`

Reply:
Nvm this comment looks like you do this in `sendNotifications` so you might not need this pushUnreadCount. When client receives a new post via `notifyNew` it will increase the counter for unreads.

Reply:
Using it here at line #87 actually. But this is for keeping "unread" update to date. topics.pushUnreadCount does the checks too. Maybe should use that "batch.processArray" from there though.

Reply:
Just saw your latter comment now. It indeed increases, I missed that. Going to remove.